dvc/fs/s3.py

```diff
@@ -103,9 +105,28 @@ def _prepare_credentials(self, **config):
 
         # encryptions
         additional = login_info["s3_additional_kwargs"]
-        additional["ServerSideEncryption"] = config.get("sse")
-        additional["SSEKMSKeyId"] = config.get("sse_kms_key_id")
+        sse_customer_key = None
+        if config.get("sse_customer_key"):
+            if config.get("sse_kms_key_id"):
+                raise ConfigError(
+                    "`sse_kms_key_id` and `sse_customer_key` AWS S3 config "
+                    "options are mutually exclusive"
+                )
+            import base64
+
+            sse_customer_key = base64.b64decode(config.get("sse_customer_key"))
+        sse_customer_algorithm = config.get("sse_customer_algorithm")
+        if not sse_customer_algorithm and sse_customer_key:
+            sse_customer_algorithm = "AES256"
+        sse_params = SSEParams(
+            server_side_encryption=config.get("sse"),
+            sse_customer_algorithm=sse_customer_algorithm,
+            sse_customer_key=sse_customer_key,
+            sse_kms_key_id=config.get("sse_kms_key_id"),
+        )
```

Comment:
I wonder if these belong in s3fs?

Reply:
Not sure exactly what the question meant. However, I would go ahead explain the snippet tagged. This `SSEParams` is just a helper class which accepts these parameters and provides a method `to_kwargs()` which  returns a dictionary with appropriate key names required by aws libraries. Before these changes, the key names were hardcoded like below
```
        additional["ServerSideEncryption"] = config.get("sse")
        additional["SSEKMSKeyId"] = config.get("sse_kms_key_id")
```
Now, this is achieved by appropriately constructing the SSEParams object, and then updating the `additional` dictionary with the dictionary obtained from `to_kwargs()` like below

```
        additional.update(sse_params.to_kwargs())

```